contracts/access/sources/ownership_transfer/delayed.move

```diff
@@ -0,0 +1,232 @@
+    pending: Option<PendingTransfer>,
+}
+
+/// Snapshot of a scheduled transfer or unwrap, including the execution time.
+/// A non-existing recipient means an unwrap is scheduled.
+public struct PendingTransfer has drop, store {
+    recipient: Option<address>,
+    execute_after_ms: u64,
+}
+
+/// Hot potato to ensure a wrapped object was returned after being taken using
+/// the `borrow_val` call.
+public struct Borrow { wrapper_id: ID, object_id: ID }
+
+// === Events ===
+
+/// Emitted when a delayed transfer is scheduled.
+public struct TransferScheduled has copy, drop {
+    wrapper_id: ID,
+    current_owner: address,
+    new_owner: address,
+    execute_after_ms: u64,
+}
+
+/// Emitted when an unwrap is scheduled.
+public struct UnwrapScheduled has copy, drop {
+    wrapper_id: ID,
+    current_owner: address,
+    execute_after_ms: u64,
+}
+
+/// Emitted when a delayed transfer is executed.
+public struct OwnershipTransferred has copy, drop {
+    wrapper_id: ID,
+    previous_owner: address,
+    new_owner: address,
+}
+
+// === Wrap / unwrap / borrow ===
+
+/// Wrap a capability/object in a delayed transfer wrapper with the desired minimum delay. The
+/// capability is tucked under a dynamic field so its object ID remains discoverable.
+public fun wrap<T: key + store>(
+    cap: T,
+    min_delay_ms: u64,
+    ctx: &mut TxContext,
+): DelayedTransferWrapper<T> {
+    let mut wrapper = DelayedTransferWrapper {
+        id: object::new(ctx),
+        min_delay_ms,
+        pending: option::none(),
+    };
+    dof::add(&mut wrapper.id, Wrapped {}, cap);
+    wrapper
+}
+
+/// Borrow the wrapped capability immutably—useful for inspection without touching the schedule.
+public fun borrow<T: key + store>(self: &DelayedTransferWrapper<T>): &T {
+    dof::borrow(&self.id, Wrapped {})
+}
+
+/// Borrow the wrapped capability mutably when internal state needs to be tweaked without editing
+/// the pending schedule.
+public fun borrow_mut<T: key + store>(self: &mut DelayedTransferWrapper<T>): &mut T {
+    dof::borrow_mut(&mut self.id, Wrapped {})
+}
+
+/// Take the wrapped capability from the `DelayedTransferWrapper` with a guarantee that it will be returned.
+public fun borrow_val<T: key + store>(self: &mut DelayedTransferWrapper<T>): (T, Borrow) {
+    let cap = dof::remove(&mut self.id, Wrapped {});
+    let object_id = object::id(&cap);
+    (cap, Borrow { wrapper_id: object::id(self), object_id })
+}
+
+/// Return the borrowed capability to the `DelayedTransferWrapper`. This method cannot be avoided
+/// if `borrow_val` is used.
+public fun return_val<T: key + store>(
+    self: &mut DelayedTransferWrapper<T>,
+    capability: T,
+    borrow: Borrow,
+) {
+    let Borrow { wrapper_id, object_id } = borrow;
+
+    assert!(object::id(self) == wrapper_id, EWrongDelayedTransferWrapper);
+    assert!(object::id(&capability) == object_id, EWrongDelayedTransferObject);
+
+    dof::add(&mut self.id, Wrapped {}, capability);
```

Comment:
I have to check BCS representation in deep for this cases, but I wonder if using an empty struct here is better than a vector with 5 a 5 bytes word for example? cc @damirka